README.md
# GCSnake
### A snake game build with Game Closure SDK
Version 0.2.2

## Overview

[Game Closure SDK](http://gameclosure.com) is a very promising HTML5 game development framework laucnhed to the public around Feb. 2013. To play with this new SDK, I wrote a snake game, which gradually evolved in order to be part of presentation about Game Closure.

This game doesn't really focus on the indepth capabilities of the SDK as much as getting the feel of writing classes and passing data around. 

## Setup:

If you haven't already, go grab your copy of the [Game Closure SDK](http://docs.gameclosure.com/guide/install.html), install it, and make sure that it's working by running `basil serve`.

#### Steps:
+ Download/Clone the GCSnake project form GitHub.
+ Move the project within the devkit directory, if it is outside.
+ In terminal, navigate to the game's directory and execute `basil register` to add the game to your registry.
+ Run the game as you would with all other games you created using Game Closure.

## Screenshots

Simple countdown to enhance UX<br />
![screen_1](http://i1129.photobucket.com/albums/m509/mazyod/download4.png)

Snake game with wrapping<br />
![screen_2](http://i1129.photobucket.com/albums/m509/mazyod/download1.png)

Losing the game, with reload option<br />
![screen_3](http://i1129.photobucket.com/albums/m509/mazyod/download-2.png)

RELEASE-NOTES.md
# GDGSnake Release Notes:

v0.1.0
---
+ initial release
+ stable, but a lot needs to be done

v0.2.0
---
+ NEW: Score tracking
+ NEW: Lose functionality
+ NEW: Reset game functionality

v0.2.1
---
* FIX: Reset game crash

v0.2.2
---
* UPDATE: Read me


src/GameLogic.js

import event.Emitter as Emitter;
import lib.Enum as Enum;

/*
	CLASS DESCRIPTION:
	==================
		This class runs the game from a logical perspective. The snake
	movement, eat, growth, and collision with itself are all calculated
	in this class. The calculations are done each 'tick' and emitted to
	the GameScene in order to update the UI accordingly.

		The game is represented as a matrix, where the location of the
	game elements can be calculated using a pair of numbers (row, column)
	where row = [0, rows), column = [0, columns].

*/


/* ---- LOGIC CONSTANTS ---- */

// GAME BOARD SIZE:
var rows = 21; // should be a multiple of 3
// because we have aspect ratio of 4:3
var columns = Math.round(rows * 4.0/3.0);

// INITIAL SNAKE SETUP:
var snakeStartLength = 4;
var snakeSpeed = 2.5;

var DIRECTION = Enum({
	UP: 0, 
	DN: 3, 
	LF: 1, 
	RT: 2
});

exports = Class( Emitter, function (supr) {
	// the designated constructor
	this.init = function () {
		// initialize everything, actually
		this._initializeGameBoard();
	}

	// to make the game logic start ticking, call this function
	this.startGame = function () {
		this._tickID = setInterval(this._tick.bind(this), 500.0 / snakeSpeed);
	}

	// this method is called when the user input is captured and emitted
	this.moveSnake = function (direction) {
		// if current direction doesn't equal the opposite direction
		if (this._curDirection !== (~direction & 3)) {
			this._newDirection = direction;
		}
	}

	// resets the game logic to initial state
	this.reset = function () {
		this._initializeGameBoard();
	}

	// prepares the initial game setup
	this._initializeGameBoard = function () {
		// 1 - Prepare the game board.
		var gameBoard = new Array(rows);
		for (var i = 0 ; i < rows ; ++i) {
			gameBoard[i] = new Array(columns);
			for (var j = 0; j < columns ; ++j) {
				gameBoard[i][j] = false;
			}
		}

		// 2 - prepare the snake, both the object and on the gameboard
		var snake = [];
		for (var i = 0; i < snakeStartLength ; i++) {
			// true means snake is occupying that block
			gameBoard[0][i] = true;
			snake.push({
				r: 0,
				c: snakeStartLength - i - 1,
			});
		}

		this._gameBoard = gameBoard;
		this._snake = snake;
		this._curDirection = this._newDirection = DIRECTION.RT;
		this._tabemono = this._generateTabemono();
		this._score = 0;
	}

	// generate food that doesn't intersect with the snake's body.
	this._generateTabemono = function () {
		var tabemono;
		do {
			tabemono = {
				r: Math.floor(Math.random() * rows),
				c: Math.floor(Math.random() * columns)
			};
		} while (this._gameBoard[tabemono.r][tabemono.c]);

		return tabemono;
	}

	this._tick = function () {		
		var direction = this._curDirection = this._newDirection;
		var tabemono = this._tabemono;
		// here, we copy the snakeHead from the array, so we dont modify the original
		var snakeHead = merge(null, this._snake[0]);

		switch (direction) {
			case DIRECTION.UP:
			snakeHead.r--;
			break;

			case DIRECTION.DN:
			snakeHead.r++;
			break;

			case DIRECTION.LF:
			snakeHead.c--;
			break;

			case DIRECTION.RT:
			snakeHead.c++;
			break;

			default:
			console.error('Unkown Direction');
		}

		// wrap the head if it went outside the gameboard bounds
		(snakeHead.r < 0 ? snakeHead.r = rows - 1 : (snakeHead.r >= rows ? snakeHead.r = 0 : /* NOP */ snakeHead));
		(snakeHead.c < 0 ? snakeHead.c = columns - 1 : (snakeHead.c >= columns ? snakeHead.c = 0 : /* NOP */ snakeHead));
		// insert snakeHead at index 0... CoffeeScript 1, JavaScript 0
		this._snake.splice(0, 0, snakeHead);

		var didLose = false;
		// if the head intersects the location of the tabemono...
		var didEat = (snakeHead.r === tabemono.r && snakeHead.c === tabemono.c);
		if (didEat) {
			// ...we generate a new tabemono without popping the tail which increases the
			// snake size. There is no way the user can lose in this situation, so don't check
			this._tabemono = tabemono = this._generateTabemono();
			this._score++;
		} else {
			// else, pop the tail, and set it empty on the gameboard
			var tail = this._snake.pop();
			this._gameBoard[tail.r][tail.c] = false;
			// if the new location is already occupied, the user lost
			didLose = this._gameBoard[snakeHead.r][snakeHead.c];
		}
		// set the new location to occupied in the gameboard
		this._gameBoard[snakeHead.r][snakeHead.c] = true;
		// if we lost, stop ticking
		if (didLose) {
			clearInterval(this._tickID);
		}
		// finally, update the GameScene
		this.emit('logicUpdated', {
			didLose: didLose,
			didEat: didEat,
			newLocation: snakeHead,
			tabemono: tabemono,
			score: this._score
		});

	}

	// Simply a function that prints the gameboard nicely
	this._debugGameBoard = function () {
		var board = '';
		for (var i = 0 ; i < rows ; ++i) {
			var row = '';
			for (var j = 0; j < columns ; ++j) {
				row += (this._gameBoard[i][j] ? '■' : '□');
			}
			board += row + '\n';
		}
		console.log(board);
	}

});


/* ---- EXPORT CONSTANTS ---- */

// GAME BOARD SIZE:
// because we have aspect ratio of 4:3
exports.ROWS = rows;
exports.COLUMNS = columns;		

exports.SNAKE_START_LENGTH = snakeStartLength;
exports.SNAKE_SPEED = snakeSpeed;

exports.DIRECTION = DIRECTION;



src/Application.js
/* @license
 * This file is part of the Game Closure SDK.
 *
 * The Game Closure SDK is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * The Game Closure SDK is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with the Game Closure SDK.  If not, see <http://www.gnu.org/licenses/>.
 */

import ui.View as View;
import device;
import src.CountdownView as CountdownView;
import src.GameLogic as GameLogic;
import src.GameScene as GameScene;

exports = Class(GC.Application, function () {

	this.initUI = function () {
		// set up our object
		var gameScene = new GameScene({
			superview: this,
			layout: 'box',
			layoutWidth: '100%',
			layoutHeight: '100%'
		});

		var countdown = new CountdownView();
		var gameLogic = new GameLogic();
		// make the gameScene listen to the gamelogic's ticks
		gameScene.on('newDirection', bind(gameLogic, 'moveSnake'));
		// make the logic listen to the gamescene's user input
		gameLogic.on('logicUpdated', bind(gameScene, 'snakeMoved'));
		// make the application listen to a reset event
		gameScene.on('resetGame', bind(this, 'resetGame'));
		// start the countdown ater 1.5 sec
		setTimeout( function () {
			countdown.beginAnimation(countdownCallback.bind(this));
		}.bind(this), 1500);

		this._gameLogic = gameLogic;
		this._gameScene = gameScene;
	};

	// resets the game, both logic and scene
	this.resetGame = function () {
		this._gameLogic.reset();
		this._gameScene.reset();

		new CountdownView().beginAnimation(countdownCallback.bind(this));
	}

	// start the game as soon as the countdown finishes
	function countdownCallback () {
		this._gameLogic.startGame();
	}
	
	this.launchUI = function () {};
});


src/CountdownView.js

import ui.TextView as TextView;
import animate;

/*
	CLASS DESCRIPTION:
	==================
		This class shows a countdown on the screen. The object should be
	created, and then, beginAnimation should be called to start the animation.
	By passing a callback to the beginAnimation, you will be notified when the 
	countdown anim ends.

		I believe this class is very straight forward and ecplains itself, so I
	will refrain from commenting here.
	
*/

exports = Class( function () {
	
	this.init = function () {
		this._textViews = [
		animationTextView('3'),
		animationTextView('2'),
		animationTextView('1'),
		animationTextView('GO!')];
	};

	this.beginAnimation = function (callback) {
		for (var i = 0; i < this._textViews.length; i++) {
			var textView = this._textViews[i];
			animateView(textView, i);
		}

		setTimeout(callback, 3200);
	}
	
	function animationTextView (text) {
		return new TextView({
			superview: GC.app.view,
			layout: "box",
			text: text,
			color: "#FFF",
			layoutWidth: '60%',
			layoutHeight: '60%',
			offsetY: -30,
			size: 10000,
			centerX: true,
			centerY: true,
			inLayout: false,
			opacity: 0.0,
			centerAnchor: true,
			fontFamily: 'Serif'
		});
	}

	function animateView (view, index) {
		return animate(view).wait(index * 1000).then({
			scale: 1.4,
			opacity: 1.0
		}, 600, animate.linear).then({
			scale: 1.6
		}, 300, animate.linear).then({
			scale: 1.8,
			opacity: 0.0
		}, 300, animate.linear).then(function () {
			view.removeFromSuperview();
		});
	}

});

src/GameScene.js

import src.GameLogic as GameLogic;
import ui.View as View;
import ui.TextView as TextView;
import ui.widget.ButtonView as ButtonView;
import device;

/*
	CLASS DESCRIPTION:
	==================
		The main GUI of the game is created and added here. Currently,
	the only components are:
		1- Canvas:
			The plane where the snake can more around. In order to make 
			the size of the plane consistent on all devices, we assumed
			4:3 ratio, and all remaining space is simply black.

		2- Snake:
			An array of views that represent the snake's body. The snake
			is moved by removing the last block, and repositioning it to
			the new location of the snake.

		3- Tabemono:
			This is a word in Japanese for 'edible'. So, this represents
			the food that the snake eats in order to score and grow.

*/

/* ---- GAME SCENE CONSTANTS ---- */

var snakeColor = '#0A0';
var snakeLoseColor = '#F00';
var tabemonoColor = '#FFF';

var blockLength;

exports = Class(View, function (supr) {
	// Puts the components into place, but not game initializations
	this.init = function (opts) {
		// call super init, with 1 argument(s)
		supr(this, 'init', [opts]);
		// Setup our canvas which holds the whole game
		var canvas = new View({
			superview: this,
			// If in doubt, ALWAYS USE BOX LAYOUT
			layout: 'linear',
			centerY: true,
			centerX: true,
			// we assume 4:3 aspect ratio
			width: Math.floor((4.0/3.0) * device.screen.height),
			height: device.screen.height,
			backgroundColor: '#333333',
			justifyContent: 'end',
			direction: 'vertical'
		});

		this._canvas = canvas;

		blockLength = (canvas.style.height / GameLogic.ROWS);

		this._initializeGame();

		// register touch events for inputView
		this.on('InputStart', touchBegan.bind(this));
		this.on('InputMove', touchMoved.bind(this));
		this.on('InputSelect', touchEnded.bind(this));
		this.on('InputOut', touchEnded.bind(this));
	}

	this._initializeGame = function () {
		// Setup the scrore label
		var scoreLabel = new TextView({
			superview: this._canvas,
			layout: 'box',
			color: '#555',
			horizontalAlign: 'left',
			size: 100,
			layoutWidth: '100%',
			layoutHeight: '14%',
			padding: '10 10 10 10'
		});

		scoreLabel.updateOpts({
			text: ''
		});

		this._scoreLabel = scoreLabel;

		// the snake is basically an array of block views
		var snake = [];
		var snakeLength = GameLogic.SNAKE_START_LENGTH;
		for (var i = 0; i < snakeLength ; i++) {
			snake.push(this._newSnakeBlock(0, snakeLength - i - 1));
		};

		this._snake = snake;
		// the food representation
		var tabemono = new View({
			superview: this._canvas,
			x: -blockLength,
			y: -blockLength,
			width: blockLength,
			height: blockLength,
			backgroundColor: tabemonoColor,
			inLayout: false
		});

		this._tabemono = tabemono;
	}

	// Update the GUI after the logic changed
	this.snakeMoved = function (args) {
		var didLose = args.didLose;
		var newLocation = args.newLocation;
		var didEat = args.didEat;
		var tabemono = args.tabemono;
		var score = args.score;
		// if we ate, create a new block as a head, otherwise move the tail head
		var newHead = (didEat ? this._newSnakeBlock(0, 0) : this._snake.pop());
		newHead.updateOpts({
			x: newLocation.c * blockLength,
			y: newLocation.r * blockLength,
		});
		// insert the new head at the beginning of the array
		this._snake.splice(0, 0, newHead);
		// if we ate, or this is the first time we are running
		if (didEat || this._tabemono.style.x < 0) {
			// Move the tabemono to the Logic's new location
			this._tabemono.updateOpts({
				x: tabemono.c * blockLength,
				y: tabemono.r * blockLength
			});
			// update the score label
			this._scoreLabel.updateOpts({
				text: 'SCORE: ' + (score * 25)
			});
		} else if (didLose) {
			this._showEndGameSequence(newLocation);
		}
	}

	// reset the game
	this.reset = function () {
		// remove old stuff
		this._canvas.removeAllSubviews();
		this._initializeGame();
	}

	// Called when the game ends (ie, user loses)
	this._showEndGameSequence = function (newLocation) {
		// else if we lost, add a red block where the collision happened
		var collisionBlock = this._newSnakeBlock(newLocation.r, newLocation.c).updateOpts({
			backgroundColor: snakeLoseColor
		});
		// push it to the snake for the sake of maintaining a ref to it
		this._snake.push(collisionBlock);
		// add a tranparent button in the middle to reset
		var resetButton = new ButtonView({
			superview: this._canvas,
			title: 'Play Again',
			layout: 'box',
			inLayout: false,
			backgroundColor: '#999',
			opacity: 0.7,
			text: {
				color: '#FFFFFF',
				size: 24
			},
			width: 160,
			height: 50,
			centerX: true,
			centerY: true,
			// event handling
			on : {
				up: resetPressed.bind(this)
			}
		});

		this._resetButton = resetButton;
	}

	// create and return a view representing a block from the snake
	this._newSnakeBlock = function (row, column) {
		return new View({
			superview: this._canvas,
			inLayout: false,
			x: column * blockLength,
			y: row * blockLength,
			// IMPORTANT: Increase block size a bit to fill gaps between blocks
			width: blockLength * 1.05,
			height: blockLength * 1.05,
			backgroundColor: snakeColor
		});
	}

	/* ---- TOUCH HANDLER FUNCTIONS ---- */

	function resetPressed (event) {
		this.emit('resetGame');
	}

	var _capturedTouches = null;
	// Capture the initial point we touched upon
	function touchBegan (event, point) {
		_capturedTouches = [point];
	}

	// Capture drag event to calculate the direction
	function touchMoved (event, point) {
		// If the event didn't start within our bounds, ignore
		if (_capturedTouches === null) {
			return;
		}
		// add the new captured touch to our array
		_capturedTouches.push(point);
		// Capture 3 additional points for accuracy
		if (_capturedTouches.length > 3) {

			var totalDeltaX = 0, prevX = 0;
			var totalDeltaY = 0, prevY = 0;
			// calculate the deltas to know which direction the swipe was
			_capturedTouches.forEach(function (arrPoint) {
				if (prevX !== 0) {
					totalDeltaX += (arrPoint.x - prevX);
					totalDeltaY += (arrPoint.y - prevY);
				}

				prevX = arrPoint.x;
				prevY = arrPoint.y;
			});

			var direction;
			if (Math.abs(totalDeltaX) > Math.abs(totalDeltaY)) {
				if (totalDeltaX > 0) {
					direction = GameLogic.DIRECTION.RT;
				} else {
					direction = GameLogic.DIRECTION.LF;
				}
			} else {
				if (totalDeltaY > 0) {
					direction = GameLogic.DIRECTION.DN;
				} else {
					direction = GameLogic.DIRECTION.UP;
				}
			}
			// Finally, inform the logic of this input
			this.emit('newDirection', direction);
			// reset current touches, but continue capturing more
			_capturedTouches = [];
		}
	}

	// reset everything when the touch ends
	function touchEnded (event, point) {
		_capturedTouches = null;
	}

});